### PR TITLE
planner: support ScalarSubQuery to display them in EXPLAIN

### LINKED ISSUE #22076: subquery execution should be displayed in explain and explain analyze
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
mysql> explain select * from t where 0 = (select sleep(3));
+------------------------+---------+-----------+---------------+--------------------------------+
| id                     | estRows | task      | access object | operator info                  |
+------------------------+---------+-----------+---------------+--------------------------------+
| TableReader_12         | 1.00    | root      |               | data:TableFullScan_11          |
| └─TableFullScan_11     | 1.00    | cop[tikv] | table:t       | keep order:false, stats:pseudo |
+------------------------+---------+-----------+---------------+--------------------------------+
2 rows in set (3.00 sec)

mysql> explain analyze select * from t where 0 = (select sleep(3));
+------------------------+---------+---------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+-----------+------+
| id                     | estRows | actRows | task      | access object | execution info                                                                                                                     | operator info                  | memory    | disk |
+------------------------+---------+---------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+-----------+------+
| TableReader_12         | 1.00    | 1       | root      |               | time:205.2µs, loops:2, cop_task: {num: 1, max: 115.7µs, proc_keys: 0, rpc_num: 1, rpc_time: 83.1µs, copr_cache_hit_ratio: 0.00}    | data:TableFullScan_11          | 150 Bytes | N/A  |
| └─TableFullScan_11     | 1.00    | 1       | cop[tikv] | table:t       | tikv_task:{time:1.8µs, loops:1}                                                                                                    | keep order:false, stats:pseudo | N/A       | N/A  |
+------------------------+---------+---------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+-----------+------+
2 rows in set (3.00 sec)
```

### 2. What did you expect to see? (Required)

1. **explain**: 
    - The subquery execution plan should be displayed in the explain results
    - The explain query should not run for 3 seconds
    - The explain query should not really execute the subquery.
2. **explain analyze**: 
    - The subquery execution plan should be displayed in the explain analyze results
    - The subquery execution time (3 seconds) should be recorded in the explain analyze results.

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close #22076

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/casetest/physicalplantest/physical_plan_test.go [hunk 1]
```diff
@@ -2589,3 +2589,71 @@ func TestIndexMergeOrderPushDown(t *testing.T) {
 		tk.MustQuery("show warnings").Check(testkit.Rows(output[i].Warning...))
 	}
 }
+
+func TestExplainNonEvaledSubquery(t *testing.T) {
```
Comment (hawkingrei) [inline planner/core/casetest/physicalplantest/physical_plan_test.go:2593]: ```physicalplantest``` also has too many tests. so I suggest you create a new package to put the case test of the scalarSubQuery.

File: build/BUILD.bazel [hunk 2]
```diff
@@ -173,7 +173,6 @@ nogo(
            }) +
            select({
                "//build:without_rbe": [
-                   "//build/linter/filepermission",
```
Comment (elsa0520) [inline build/BUILD.bazel:176]: Why this line is removed?

File: sessionctx/variable/tidb_vars.go [hunk 3]
```diff
@@ -1258,6 +1260,7 @@ const (
 	DefTiDBSkewDistinctAgg                         = false
 	DefTiDB3StageDistinctAgg                       = true
 	DefTiDB3StageMultiDistinctAgg                  = false
+	DefTiDBOptExplainNoEvaledSubquery              = false
```
Comment (elsa0520) [inline sessionctx/variable/tidb_vars.go:1263]: Prefer `DefTiDBOptExplainEvaledSubquery = true`.
Although they mean the same thing. But the double negation of ```no xxx = false``` is more difficult to understand.

File: executor/executor.go [hunk 4]
```diff
@@ -2082,7 +2082,7 @@ func ResetContextOfStmt(ctx sessionctx.Context, s ast.StmtNode) (err error) {
 	sc.OriginalSQL = s.Text()
 	if explainStmt, ok := s.(*ast.ExplainStmt); ok {
 		sc.InExplainStmt = true
-		sc.ExplainFormat = explainStmt.Format
```
Comment (elsa0520) [inline executor/executor.go:2085]: Maybe we need to keep ```ExplainFormat``` together ?

File: planner/core/scalar_subq_expression.go [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+	return &ret
+}
+
+// Equal implements the Expression interface.
+func (s *ScalarSubQueryExpr) Equal(_ sessionctx.Context, e expression.Expression) bool {
+	anotherS, ok := e.(*ScalarSubQueryExpr)
+	if !ok {
+		return false
+	}
+	if s.ScalarSubQuery == anotherS.ScalarSubQuery {
```
Comment (elsa0520) [inline planner/core/scalar_subq_expression.go:139]: Maybe check ```scalarSubQueryID``` is more better ?

File: util/plancodec/id.go [hunk 6]
```diff
@@ -137,6 +137,8 @@ const (
 	TypeImportInto = "ImportInto"
 	// TypeSequence is the type of Sequence
 	TypeSequence = "Sequence"
+	// TypeScalarSubQuery is the type of ScalarQuery
+	TypeScalarSubQuery = "ScalarQuery"
```
Comment (elsa0520) [inline util/plancodec/id.go:141]:
```suggestion
	TypeScalarSubQuery = "ScalarSubQuery"
```

File: planner/core/expression_rewriter.go [hunk 7]
```diff
@@ -1086,6 +1108,19 @@ func (er *expressionRewriter) handleScalarSubquery(ctx context.Context, v *ast.S
 		er.err = err
 		return v, true
 	}
+	if er.b.ctx.GetSessionVars().StmtCtx.InExplainStmt && !er.b.ctx.GetSessionVars().StmtCtx.InExplainAnalyzeStmt && er.b.ctx.GetSessionVars().ExplainNonEvaledSubQuery {
+		scalarSubQ := ScalarSubQueryExpr{
```
[the remark this answers is not in the document]
Comment (winoros) [reply planner/core/expression_rewriter.go:1112]: I'll fix it in the later commits.

File: executor/executor.go (hunk 4 above)
Comment (winoros) [reply executor/executor.go:2085]: Yeah, it's a mistake

File: planner/core/scalar_subq_expression.go [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/codec"
+)
+
+// ScalarSubQueryExpr is a expression placeholder for the non-correlated scalar subqueries which can be evaluated during optimizing phase.
+type ScalarSubQueryExpr struct {
```
[the remark this answers is not in the document]
Comment (winoros) [reply planner/core/scalar_subq_expression.go:33]: Decoupling the codes in `core` is not easy work...
And you'll see that the struct embeds `basePlan` in it, which means we cannot move it out from the package...

I have tried to decouple some codes in `core` before. But it's really a time costing work.

File: planner/core/scalar_subq_expression.go (hunk 5 above)
Comment (winoros) [reply planner/core/scalar_subq_expression.go:139]: Changed

File: sessionctx/variable/session.go [hunk 9]
```diff
@@ -744,6 +744,9 @@ type SessionVars struct {
 	// PlanColumnID is the unique id for column when building plan.
 	PlanColumnID atomic.Int64
 
+	// MapScalarSubQ maps the scalar sub queries from its ID to its struct.
+	MapScalarSubQ map[int64]interface{}
```
[the remark this answers is not in the document]
Comment (winoros) [reply sessionctx/variable/session.go:748]: changed

File: planner/core/expression_rewriter.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (winoros) [reply planner/core/expression_rewriter.go:1112]: fixed

**Review by elsa0520 — APPROVED** [review]
LGTM

File: planner/core/scalar_subq_expression.go [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+	err := s.selfEvaluate()
+	if err != nil {
+		return s.Value, err
+	}
+	return s.Value, nil
+}
+
+// EvalInt returns the int64 representation of expression.
+func (*ScalarSubQueryExpr) EvalInt(_ sessionctx.Context, _ chunk.Row) (val int64, isNull bool, err error) {
+	return 0, false, errors.Errorf("Evaluation methods is not implemented for ScalarSubQueryExpr")
```
[the remark this answers is not in the document]
Comment (chrysan) [reply planner/core/scalar_subq_expression.go:122]: Need to revisit and implement those EvalXXX later, when we need to support execution under ScalarSubqueryExpr. Please add some comments here.

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

File: planner/core/casetest/scalarsubquery/testdata/plan_suite_out.json [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        ],
+        "Error": ""
+      },
+      {
+        "SQL": "explain format = 'brief' select * from t1 where exists(select 1 from t2 where a = (select a from t3 limit 1) and b = a);",
+        "Plan": [
+          "Selection 8000.00 root  ScalarQueryCol#15",
+          "└─TableReader 10000.00 root  data:TableFullScan",
+          "  └─TableFullScan 10000.00 cop[tikv] table:t1 keep order:false, stats:pseudo",
+          "ScalarSubQuery N/A root  Output: ScalarQueryCol#13",
```
[the remark this answers is not in the document]
Comment (winoros) [reply planner/core/casetest/scalarsubquery/testdata/plan_suite_out.json:74]: We can left for future optimization

**Review by chrysan — APPROVED** [review]

**Review by AilinKid — APPROVED** [review]

**Review by AilinKid — APPROVED** [review]

**Comment by hawkingrei** [comment]
> /hold

Waiting for the configure review's approval.

**Review by qw4990 — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**